.travis.yml

```diff
@@ -3,18 +3,19 @@ dist: xenial
 sudo: false
 before_install:
 - gem install bundler
+- curl --fail --retry 3 --retry-delay 1 --connect-timeout 3 --max-time 30 https://cli-assets.heroku.com/install.sh | sh
 - composer self-update
 install:
 - bundle install
 before_script:
 - bundle exec hatchet ci:setup
-- heroku update
-script: bundle exec parallel_rspec -n 7 --group-by runtime --unknown-runtime 1 --runtime-log "test/var/log/parallel_runtime_rspec.${STACK}.log" --verbose-process-command --combine-stderr --prefix-output-with-test-env-number test/spec/
+script: bundle exec parallel_rspec -n 16 --group-by runtime --unknown-runtime 1 --runtime-log "test/var/log/parallel_runtime_rspec.${STACK}.log" --verbose-process-command --combine-stderr --prefix-output-with-test-env-number test/spec/
```

Comment:
That's correct.

The reason is that a bunch of test files, using filters, are skipped entirely for certain stacks, or when a PHP version series isn't available on the stack.

For instance, PHP 5.5 is only available on cedar-14, and PHP 7.4 is not available on cedar-14.

So for these test case files, no tests are executed, and parallel_rspec will not produce output in the runtime log (which for convenient copy/paste for periodical updating in Git purposes gets `cat`ed at the end of the run), meaning it can't know how long these take, and would then assume an average of all known runtimes (IIRC).

I used to manually fake a value of 1 second in the runtime log, but this is easier, as you can't miss a file.

In effect, it results in these skipped test files not taking up any "space" in the parallel balancing buckets (well, one second, but that's fine).

Reply:
> If so, why 1 second and not the default (average) or something longer? Just curious for when I start working on my own Hatchet tests.

We could upstream new behavior here if desired. Right now there are two ways of running tests in multiple processes: split tests by file or split ALL tests. DZ is testing behavior where one app is deployed and then many multiple runs are happening against the same app, if we split on ALL tests then it would trigger all of them to deploy an app see https://github.com/grosser/parallel_split_test#beforeall-rspec-hooks.

I think that the python tests uses the split ALL tests strategy.

Another possibility is to get a license for the knapsack lib https://github.com/KnapsackPro/knapsack_pro-ruby as it seems to store test run data automatically and not require manually syncing files. Another 3rd option would be to investigate moving somewhere with a semi-durable store between runs. We could maybe switch over to CircleCI and then use the cache to track values between executions.

It seems like you've got this specific question covered, i'm just letting yall know about a few more possibilities. Most notably: I'm happy to do upstream work on other gems/libs that we depend on if owe find something we think would be worthwhile.

Reply:
@schneems Thank you - that's filled in a few gaps in my understanding, very helpful :-D